src/Simulation/Simulators/SparseSimulator/SparseSimQSharpTests/Program.qs

```diff
@@ -0,0 +1,643 @@
+    operation LargeStateTests() : Unit {
+        let nqubits = 12;
+        LargeStateTestWrapper(Rotation1CompareTest, nqubits);
+        LargeStateTestWrapper(RotationCompareTest, nqubits);
+        LargeStateTestWrapper(SWAPTest, nqubits);
+        LargeStateTestWrapper(CSWAPTest, nqubits);
+        LargeStateTestWrapper(CNOTTest, nqubits);
+        LargeStateTestWrapper(ResetTest, nqubits);
+        LargeStateTestWrapper(AssertTest, nqubits);
+        LargeStateTestWrapper(AndChainTest, nqubits);
+        LargeStateTestWrapper(CZTest, nqubits);
+        LargeStateTestWrapper(AllocationTest, nqubits);
+        LargeStateTestWrapper(Rotation1CompareTest, nqubits);
+        LargeStateTestWrapper(RotationFracCompareTest, nqubits);
+
+    }
+
+    operation _ToffoliCSwap(targets : Qubit[]) : Unit is Ctl + Adj {
+        CCNOT(targets[0], targets[1], targets[2]);
+        CCNOT(targets[0], targets[2], targets[1]);
+        CCNOT(targets[0], targets[1], targets[2]);
+    }
+    operation _ArrayCSwap(targets : Qubit[]) : Unit is Ctl + Adj {
+        (Controlled SWAP)([targets[0]], (targets[1], targets[2]));
+    }
+    @Test("Microsoft.Quantum.SparseSimulation.SparseSimulator")
+    operation CCNOTvToffoliTest() : Unit {
+        AssertOperationsEqualReferenced(3, _ArrayCSwap, _ToffoliCSwap);
+    }
+
+    // Creates a big state so an existing test forces parallel execution
+    operation LargeStateTestWrapper(test : (Unit => Unit), nqubits : Int) : Unit {
+        let prob = PowD(0.5, IntAsDouble(nqubits));
+         use qubits = Qubit[nqubits];
+        for idx in 0..nqubits - 1 {
+            H(qubits[idx]);
+        }
+        for idy in 1..128 {
+            CCNOT(qubits[idy % nqubits], qubits[(idy+1) % nqubits], qubits[(idy+2) % nqubits]);
+        }
+        test();
+        ResetAll(qubits);
+    }
+
+    operation MCXTime(nqubits : Int, ngates : Int) : Unit {
+        use qubits = Qubit[nqubits];
+        for idx in 0..nqubits - 1 {
+            H(qubits[idx]);
+        }
+        for idx in 0..nqubits - 1 {
+            CNOT(qubits[idx], qubits[(idx+1)% nqubits]);
+        }
+        for idy in 1..ngates {
+            CNOT(qubits[idy %nqubits], qubits[(idy+1)%nqubits]);
+        }
+        ResetAll(qubits);
+    }
+
+    @Test("Microsoft.Quantum.SparseSimulation.SparseSimulator")
+	operation PartialDumpTest() : Unit  {
+        use qubits = Qubit[4] {
+            ApplyToEach(H, qubits);
+            CNOT(qubits[2], qubits[3]);
+            DumpRegister("Test_file_1", qubits[0..1]);
+            DumpRegister("Test_file_2", qubits[2..3]);
+            CNOT(qubits[2], qubits[3]);
+            H(qubits[2]);
+            CNOT(qubits[0], qubits[2]);
+            DumpRegister("Test_file_3", qubits[0..1]);
+            DumpRegister("Test_file_4", qubits);
+            ResetAll(qubits);
+        }
+    }
+
+
+    operation _FakeR1Frac(numerator : Int, denominator : Int, qubit : Qubit[]) : Unit is Adj + Ctl {
```

Comment:
It does not call the R1 function internally, but simulates the effect with two R rotations. The idea is to test that the two types of rotation are the same.

Reply:
Understood, and I see you're using it to call the `AssertOperationsEqualReferenced` diagnostic operation, makes sense. In this case, I wouldn't think of this as "fake," then, I guess, but a decomposition of the `R1Frac` operation into `RFrac` calls. That said, since the `RFrac(PauliI, _, _, _)` call doesn't do anything other than impact a global phase, I'm not sure that as written this tests that the global phase imparted is correct?

Reply:
That's all true. The sparse simulator doesn't have a well-defined global phase anyway, so it shouldn't be able to check global phase. It would probably be more useful if it also ran a test of controlled versions of the same gates.